architecture/sql-layer.md

```diff
@@ -0,0 +1,89 @@
+
+#### Parsing
+
+CockroachDB's supported SQL API is described in a [`yacc`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/parser/sql.y) file, which we've derived from PostgreSQL. The received query is then parsed against the `yacc` file, converting the string version of the query into [Abstract Syntax Trees (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree).
+
+#### Planning
+
+Given the AST from the parser, CockroachDB begins planning the query's execution by generating a tree of `planNodes`.
+this is ultimately how CockroachDB converts SQL statements into KV operations. Each of these nodes is a contained set
+of code that can be executed to operate on the data from the KV layer.
+
+This step also includes steps analyzing the client's SQL statements against the expected AST expressions, which includes things like type checking.
+
+You can see the planNodes a query generates using [`EXPLAIN`](/docs/explain.html).
+
+#### Executing
+
+planNodes are then executed, which begins by communicating with the Transaction Layer.
+
+This step also includes encoding values from your statements, as well as decoding values retrieved from the KV layer.
+
+### Encoding
+
+Though SQL queries are written in parsable strings, lower layers of CockroachDB deal primarily in bytes. This means at the SQL layer, in query execution, CockroachDB must convert row data from their SQL representation as strings into bytes, and convert bytes returned from lower layers into SQL data that can be passed back to the client.
+
+It's also important––for indexed columns––that this byte encoding preserve the same sort order as the data type it represents. This is because of the way CockroachDB ultimately stores data in a sorted key-value map; storing bytes in the same order as the data it represents lets us efficiently scan data in the key-value layer.
+
+However, for non-indexed columns (i.e. those that are covered or stored by an index), CockroachDB instead uses an encoding (known as "value encoding") which consumes less space but does not preserve ordering.
+
+You can find more exhaustive detail in the [Encoding Tech Note](https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/encoding.md).
+
+### DistSQL
+
+Because CockroachDB is a distributed database, we've begun implementing a Distributed SQL (DistSQL) optimization tool for some queries. Though its architecture is worthy of its own documentation, here's a brief outline of how it works.
+
+For a subset of SQL statements, our execution phase works slightly differently. Instead of querying each node for rows that match a query and aggregating the results on the coordinating gateway node, we bring the query to each node involved. Each node then processes the query for the rows it stores. The processed results (instead of all involved rows) are then returned to the coordinating node, which aggregates the results and returns it to the client.
```

Comment:
> we bring the query to each node involved

Maybe a couple words to say why we do that? To parallelize queries and operate on bigger datasets than would be possible to manage on one node.

Reply:
Done.